pkg/runtime/interfaces.go

```diff
@@ -23,64 +23,74 @@ import (
 	"k8s.io/kubernetes/pkg/api/unversioned"
 )
 
-// Codec defines methods for serializing and deserializing API objects.
-type Codec interface {
-	Decoder
-	Encoder
-}
+const (
+	APIVersionInternal    = "__internal"
+	APIVersionUnversioned = "__unversioned"
+)
 
-// Decoder defines methods for deserializing API objects into a given type
-type Decoder interface {
-	Decode(data []byte) (Object, error)
-	// TODO: Remove this method?
-	DecodeToVersion(data []byte, groupVersion unversioned.GroupVersion) (Object, error)
-	DecodeInto(data []byte, obj Object) error
-	// TODO: Remove this method?
-	DecodeIntoWithSpecifiedVersionKind(data []byte, obj Object, groupVersionKind unversioned.GroupVersionKind) error
-
-	DecodeParametersInto(parameters url.Values, obj Object) error
+// Typer retrieves information about an object's group, version, and kind.
+type Typer interface {
+	// ObjectKind returns the version and kind of the provided object, or an
+	// error if the object is not recognized (IsNotRegisteredError will return true).
+	// It returns whether the object is considered unversioned at the same time.
+	ObjectKind(Object) (*unversioned.GroupVersionKind, bool, error)
 }
 
-// Encoder defines methods for serializing API objects into bytes
 type Encoder interface {
-	Encode(obj Object) (data []byte, err error)
-	EncodeToStream(obj Object, stream io.Writer) error
+	// EncodeToStream writes an object to a stream. Override versions may be provided for each group
+	// that enforce a certain versioning. Implementations may return errors if the versions are incompatible,
+	// or if no conversion is defined.
+	EncodeToStream(obj Object, stream io.Writer, overrides ...unversioned.GroupVersion) error
+}
 
-	// TODO: Add method for processing url parameters.
-	// EncodeParameters(obj Object) (url.Values, error)
+type Decoder interface {
+	// Decode attempts to deserialize the provided data using either the innate typing of the scheme or the
+	// default kind, group, and version provided. It returns a decoded object as well as the actual kind, group, and
+	// version decoded, or an error. If into is non-nil, it will be used as the target type and implementations
+	// may choose to use it rather than reallocating an object. However, the object is not guaranteed to be
+	// populated. The returned object type is not guaranteed to match into.
+	Decode(data []byte, defaults *unversioned.GroupVersionKind, into Object) (Object, *unversioned.GroupVersionKind, error)
```

Comment:
Into is pretty widely used in our codebase.  It's actually more correct in
some cases than version, because internal clients ultimately need a
concrete struct to deal with, and most decodes end up with a cast.

I'm debating between two approaches to layering conversion and
serialization.  The first, in this pr, wraps a naive serializer (limited
version awareness) with a version aware codec.  The challenge is that the
interface isn't quite identical.  On the plus side, it was possible to
prove that we don't need 6 decode methods in order to satisfy our use
cases, which was something I wasn't sure about until I went and touched all
the code again to remember.  On the negative side, negotiated versions are
harder because a lot of code that has access to a codec may want to choose
the target serialization (specifically RESTMapper used by client commands
and printers).  We also need access in a few cases to the versioned object
and the original (such as in the apply code) but passing two codecs would
be silly

The alternative is an interface like Decode([]byte, DecodeOptions{}), where
DecodeOptions may optionally specify things like:

* version defaulting
* desired target version
* desired target object (into)
* expected (or allowed) serialization formats
* callback to get access to the raw deserialized object
* callback to check the type of the object prior to final conversion, or to
perform advanced decoding
* callback to transform nested encoded objects (List, Watch, etc)

I may try that interface to see whether it simplifies the code.  Same for
Encoding as well.

Ultimately the decoupling between conversion and serialization is necessary
for sanity of code, even a temporary state along the way is an improvement
- there was lots of accumulated edge cases that deserved to be flensed and
given proper solutions.


On Dec 11, 2015, at 9:00 PM, Chao Xu <notifications@github.com> wrote:

In pkg/runtime/interfaces.go
<https://github.com/kubernetes/kubernetes/pull/17922#discussion_r47424917>:

>
> -	// TODO: Add method for processing url parameters.
> -	// EncodeParameters(obj Object) (url.Values, error)
> +type Decoder interface {
> +	// Decode attempts to deserialize the provided data using either the innate typing of the scheme or the
> +	// default kind, group, and version provided. It returns a decoded object as well as the actual kind, group, and
> +	// version decoded, or an error. If into is non-nil, it will be used as the target type and implementations
> +	// may choose to use it rather than reallocating an object. However, the object is not guaranteed to be
> +	// populated. The returned object type is not guaranteed to match into.
> +	Decode(data []byte, defaults *unversioned.GroupVersionKind, into Object) (Object, *unversioned.GroupVersionKind, error)

Is it possible to remove into and the return value
*unversioned.GroupVersionKind from the signature? Then this method is pure
serialization. It will make the refactoring much more difficult, but it
seems to worth the effort.

The returned GVK is useful for type enforcement and is required in the
apiserver.  Simpler clients can just ignore it - I don't think creating
multiple methods should be necessary for that.  The resulting refactored
decode methods are notable mostly for the amount of type defaulting that is
needed.